morphem/Feature_Extraction.py

```diff
@@ -0,0 +1,142 @@
+
+    return dataset
+
+
+
+class ConvNextClass():
+    def convnext_model(gpu):
+        device = torch.device(f"cuda:{gpu}" if torch.cuda.is_available() else "cpu")
+
+      
+        feature_file = 'pretrained_convnext_channel_replicate.npy'
+     
+
+        pretrained = True
+        model = create_model("convnext_tiny.fb_in22k", pretrained=pretrained).to(device)
+        feature_extractor = nn.Sequential(
+                            model.stem,
+                            model.stages[0],
+                            model.stages[1],
+                            model.stages[2].downsample,
+                            *[model.stages[2].blocks[i] for i in range(9)],
+                            model.stages[3].downsample,
+                            *[model.stages[3].blocks[i] for i in range(3)],
+                        )
+
+        model_check = "convnext"
+        return feature_extractor, device, feature_file
+
+
+    def forward(x: torch.Tensor, gpu: int) -> torch.Tensor:
+        feature_extractor, _,  = ConvNextClass.convnext_model(gpu)
+        x = feature_extractor(x)
+        x = F.adaptive_avg_pool2d(x, (1, 1))
+      
+        return x
+
+
+class ResNetClass():
+    def resnet_model(gpu):
+        device = torch.device(f"cuda:{gpu}" if torch.cuda.is_available() else "cpu")
+
+
+        feature_file = 'pretrained_resnet18_features.npy'
+
+
+        weights = ResNet18_Weights.IMAGENET1K_V1
+        m = resnet18(weights=weights).to(device)
+        feature_extractor = torch.nn.Sequential(*list(m.children())[:-1]).to(device)
+
+        model_check = "resnet"
+
+        return weights, feature_extractor, device, feature_file
+
+
+
+def get_save_features(feature_dir, root_dir, model_check, gpu):
+    dataset_names = ['CP','Allen','HPA']
+    for dataset_name in dataset_names:
+        dataset = configure_dataset(root_dir, dataset_name)
+        train_dataloader = DataLoader(dataset, batch_size=256, shuffle=False)
+        if model_check == "resnet":
+            weights, feature_extractor, device, feature_file = ResNetClass.resnet_model(gpu)
+            preprocess = weights.transforms()
+        else:
+            _, device, feature_file = ConvNextClass.convnext_model(gpu)
+        all_feat = []
+        for images, label in tqdm(train_dataloader, total=len(train_dataloader)):
+            cloned_images = images.clone()
+            batch_feat = []
+            for i in range(cloned_images.shape[1]):
+                # Copy each channel three times 
+                channel = cloned_images[:, i, :, :]
+                channel = channel.unsqueeze(1)
+                expanded = channel.expand(-1, 3, -1, -1).to(device)
+        
+                if model_check == "resnet":
+                    expanded = preprocess(expanded).to(device)
+                    feat_temp = feature_extractor(expanded).cpu().detach().numpy()
+                else: 
+                    feat_temp = ConvNextClass.forward(expanded, gpu).cpu().detach().numpy()
```

Comment:
This line should not use the Class, it should use the model. Note that calling forward results in calling convnext_mode, which in turns creates a new model for each image, which is not efficient.